mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td

```diff
@@ -289,6 +289,9 @@ def MapOp : LinalgStructuredBase_Op<"map", [
 
     bool payloadUsesValueFromOperand(OpOperand * opOperand) {
       if (isDpsInit(opOperand)) return false;
+      if (getOperation()->getRegion(0).empty()) {
+        return true;
+      }
```

Comment:
I don't quite get this. When the region is empty (is this even allowed?), the operand is considered used? But when it has a terminator not reading from the corresponding block argument, the operand is not considered used? Why? If we assume it is read anyway, it should be marked as such in both cases.

Reply:
> I don't quite get this. When the region is empty (is this even allowed?), the operand is considered used? But when it has a terminator not reading from the corresponding block argument, the operand is not considered used? Why? If we assume it is read anyway, it should be marked as such in both cases.

Thank you for your suggestion! For this issue, here is my understanding: When region is empty, it usually occurs in scenarios where region does not need to interpret the semantics of op, such as linalg.map:

```
     %mapped = linalg.map { arith.addf }                                              
              ins(%arg0, %arg1 : tensor<10x100xf32>, tensor<10x100xf32>)           
              outs(%map_init : tensor<10x100xf32>)
```

In the example above, the linalg.map has an empty region. This is because, with arith.add present, we do not need to provide additional interpretation for this op, as its implementation simply involves adding the two input operands. In such scenarios where no interpretation is required, we consider all operands to be used in accordance with the semantics of the op.

In linalg operations implemented with regions, it is evident that when the block arguments corresponding to operands are not used, it indicates that they are not being read.

Reply:
`linalg.map { arith.addf }` is just a pretty-printed version. It has a region with an `arith.addf` and `linalg.yield` terminator.

Reply:
> `linalg.map { arith.addf }` is just a pretty-printed version. It has a region with an `arith.addf` and `linalg.yield` terminator.

You are right. I reviewed why I thought this way: before this, I didn't add handling for when region was empty, leading to some core dump issues during part of the bufferization tests. This happened because the 'cloneWithoutRegions' function was called during the bufferization process, and in this function, the side effects of 'op' were queried, unfortunately, at this time region was still empty. So, I'd like to solicit opinions from both of you: in this situation, can we conservatively assume that all operands can be read and written to?  @matthias-springer @ftynse

Reply:
Hi @matthias-springer @ftynse,I wanted to send a friendly reminder to re-review the proposal above at your earliest convenience and hope to hear your thoughts on this issue.

Reply:
Why does this lead to a crash? The bufferization impl for LinalgOps calls `cloneWithoutRegions`, but inlines the original region immediately afterwards. There should be no call to `payloadUsesValueFromOperand` in-between. (Even if there is one, this function would just return `false` and should not crash.)

Basically, the op is invalid without a region body. Generally speaking, helper functions on an op are allowed to function incorrectly if the op is invalid.

Reply:
> Why does this lead to a crash?

When the cloneWithoutRegions function is called, the bufferize rewriter will invoke the notifyOperationInserted function. This function will query the side effects of the operation to count the number of memory allocations. Because this patch will call payloadUsesValueFromOperand when querying side effects, and the payloadUsesValueFromOperand of the map operation is implemented as follows:

```
bool payloadUsesValueFromOperand(OpOperand * opOperand) {                                                                                                                                  
   if (isDpsInit(opOperand)) return false;                                                                                                                                                                                                                                                                                                                     
  return !getMatchingBlockArgument(opOperand).use_empty();
}
```
This will fetch the block inside the op's region, causing a core dump.  

> Basically, the op is invalid without a region body. Generally speaking, helper functions on an op are allowed to function incorrectly if the op is invalid.

So should we prohibit querying side effects before the block is inserted?

Reply:
Oh, I see. The op is already inserted even though it was not fully constructed yet (region missing). Ideally, we would change the `BufferizableOpInterface` impl., so that we insert the op only when it is fully constructed.

I think it's important to keep the `LinalgOp` interface clean and we should have to account for edge cases in the bufferization there.

Could you change the implementation in `Linalg/Transforms/BufferizableOpInterfaceImpl.cpp` such that it clones the op without regions without a builder. Basically using `Operation::create(OpState)` (instead of `cloneOpWithoutRegions`). Then inline the region. Then `rewriter.insert`. Then the callback in the bufferization will be triggered on a fully constructed and valid op.

Reply:
I agree with your perspective, we should keep the implementation within the interface clean. I have already modified the code according to your comment, thanks for your feedback.

Reply:
Hi @ftynse, do you have any further comments on the modifications proposed in this patch?